core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -205,15 +207,38 @@ modify_launch_configuration_if_desired_occupancy_is_specified(
 
   if (dynamic_shmem > shmem) {
     shmem        = dynamic_shmem;
-    prefer_shmem = false;
+    prefer_shmem = 0;
   }
 }
 
 template <class Policy>
 std::enable_if_t<!Policy::experimental_contains_desired_occupancy>
 modify_launch_configuration_if_desired_occupancy_is_specified(
-    Policy const&, cudaDeviceProp const&, cudaFuncAttributes const&,
-    dim3 const& /*block*/, int& /*shmem*/, bool& /*prefer_shmem*/) {}
+    Policy const&, cudaDeviceProp const& properties,
+    cudaFuncAttributes const& attributes, dim3 const& block, int& shmem,
+    int& prefer_shmem) {
+  // prefer_shmem = 0 - cudaFuncCachePreferL1
+  // prefer_shmem = 1 - cudaFuncCachePreferEqual
+  // prefer_shmem = 2 - cudaFuncCachePreferShared
+
+  // If prefer_shmem is already set to 0, return.
+  if (prefer_shmem == 0) return;
+
+  int const block_size = block.x * block.y * block.z;
+  int active_blocks    = properties.maxThreadsPerMultiProcessor / block_size;
```

Comment:
this is off. This needs to take into account the register pressure

Reply:
otherwise we gonna over estimate the shared requirement always.

Reply:
Added logic to consider the register pressure too.